BEIMA.Backend/MongoService/MongoConnector.cs

```diff
@@ -188,5 +189,134 @@ public BsonDocument UpdateDevice(BsonDocument doc)
             }
         }
+
+        /// <summary>
+        /// Gets a device type from the "deviceTypes" collection, given an objectID.
+        /// </summary>
+        /// <param name="objectId">Corresponds to the "_id" field for a given document inside of MongoDB</param>
+        /// <returns>BsonDocument that was requested</returns>
+        public BsonDocument GetDeviceType(ObjectId objectId)
+        {
+            CheckIsConnected();
+
+            var filter = Builders<BsonDocument>.Filter.Eq("_id", objectId);
+
+            try
+            {
+                var db = client.GetDatabase(dbName);
+                var devices = db.GetCollection<BsonDocument>(deviceTypeCollection);
+                return devices.Find(filter).FirstOrDefault();
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine(ex.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Gets a device from the "deviceTypes" collection, given an objectID.
+        /// </summary>
+        /// <returns>BsonDocument that was requested</returns>
+        public List<BsonDocument> GetAllDeviceTypes()
+        {
+            CheckIsConnected();
+
+            var filter = Builders<BsonDocument>.Filter.Empty;
+
+            try
+            {
+                var db = client.GetDatabase(dbName);
+                var devices = db.GetCollection<BsonDocument>(deviceTypeCollection);
+                var docs = devices.Find(filter).ToList();
+                return docs;
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine(ex.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Inserts a device into the "deviceTypes" collection
+        /// </summary>
+        /// <param name="doc">BsonDocument that contains the fully formed device type document (including all required and optional fields)</param>
+        /// <returns>ObjectId of the newly inserted object if successful, null if failed</returns>
+        public ObjectId? InsertDeviceType(BsonDocument doc)
+        {
+            CheckIsConnected();
+
+            try
+            {
+                var db = client.GetDatabase(dbName);
+                var devices = db.GetCollection<BsonDocument>(deviceTypeCollection);
+                devices.InsertOne(doc);
+                return (ObjectId)doc["_id"];
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine(ex.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Deletes from the "deviceTypes" collection, given the objectID.
+        /// </summary>
+        /// <param name="objectId">Corresponds to the "_id" field for a given document inside of MongoDB</param>
+        /// <returns>true if successful, false if not successful</returns>
+        public bool DeleteDeviceType(ObjectId objectId)
```

Comment:
Do we want to add a check to make sure that there aren't any devices associated with the given device type? Or do we want to do that somewhere else?

Reply:
I think a check would be good, then something can be passed up to frontend and a warning message can be displayed before the commit to deleting it. I think that check should be done else where though, we don't want to put too much responsibility into the Mongo Connector code

Reply:
I think the check should be placed at a higher level as well. When the user wants to delete a device type, then we should pop up a message on the front end showing how many devices they are about to delete. Only after they confirm, should we execute this function.

If we threw an exception at this low of a level, I would imagine that this function would almost always throw an exception. Probably better to do this at the endpoint level.

Reply:
Thinking about it, we may even need to do this at the frontend level, since the endpoint would just keep throwing an exception as well.